### PR TITLE
Trace processing pipeline timing

### PR DESCRIPTION
**Pull request by richardstartin**
* Renames `Monitor` (the thing we refer to as "health metrics" internally) `HealthMetrics`
* Creates a `Monitoring` class which is essentially a factory for timers and, in the future, gauges
* Implements a parsimonious thread local `Timer` based on a packed HdrHistogram
* Adds timing to various scopes int he trace processing timeline 
     - duty cycles
    -  buffer fill ups (may become a gauge in the future)
    -  agent discovery time
    - agent response time (may become a gauge in the future)
    - pending trace write lock time

### CODE REVIEW

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/PayloadDispatcherTest.groovy [hunk 1]
```diff
@@ -10,19 +11,25 @@ import datadog.trace.core.CoreTracer
...
 
 class PayloadDispatcherTest extends DDSpecification {
 
+  @Shared
+  Monitoring monitoring = new Monitoring(new NoOpStatsDClient(), 1, TimeUnit.SECONDS)
+
   def "dropped traces should be reported in the representativeCount"() {
     setup:
-    Monitor monitor = Mock(Monitor)
+    HealthMetrics monitor = Mock(HealthMetrics)
```
Comment (tylerbenson) [inline dd-trace-core/src/test/groovy/datadog/trace/common/writer/PayloadDispatcherTest.groovy:30]: I think you need a name change for `monitor` here and several other places in this PR... I won't repeat the comment over again.

**Review by tylerbenson — APPROVED** [review]
Overall looks ok.  I'm not a huge fan of the name `Monitoring` for this usage.  I consider that a nit since it's all internal and easily changed in the future.

**Comment by tylerbenson** [comment]
Maybe we should add a `Counter` type that is used by `HealthMetrics`.  I'm not sure I understand the distinction on what is a health metric and how it's distinct.  Should they be unified somehow?

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/PayloadDispatcherTest.groovy (hunk 1 above)
Comment (richardstartin) [reply dd-trace-core/src/test/groovy/datadog/trace/common/writer/PayloadDispatcherTest.groovy]: yep - groovy is a second class citizen in IDE refactoring, unless I'm missing something

**Comment by richardstartin** [comment]
> Maybe we should add a Counter type that is used by HealthMetrics. I'm not sure I understand the distinction on what is a health metric and how it's distinct. Should they be unified somehow?

If you want to model all things you would want to count or time in a domain specific way, you need to define a pretty big domain. Modeling timers as timers, you can put them where you want to time things, and don't need to elevate implementation details like "making an HTTP call" or "draining a queue" into a domain model which will change whenever you change you mind about what it makes sense to time.

File: dd-trace-core/src/main/java/datadog/trace/core/monitor/Timer.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  Timer(final String name, final String[] tags, final StatsDClient statsd, long flushAfterNanos) {
+    this.name = name;
+    this.statsd = statsd;
+    this.flushAfterNanos = flushAfterNanos;
+    this.histogram = new PackedHistogram(THIRTY_SECONDS_AS_NANOS, 3);
+    this.meanTags = mergeTags(MEAN, tags);
+    this.p50Tags = mergeTags(P_50, tags);
+    this.p99Tags = mergeTags(P_99, tags);
+    this.maxTags = mergeTags(MAX, tags);
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/monitor/Timer.java:43]: So how does this compare in the backend as opposed to a _normal_ histogram that the `agent` has computed? I would love for statsd/dogstatsd to have a way to send a histogram in a compact way as one metric.

File: dd-trace-core/src/main/java/datadog/trace/core/monitor/Recording.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package datadog.trace.core.monitor;
+
+public abstract class Recording implements AutoCloseable {
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/monitor/Recording.java:3]: This is nice. Using `AutoCloseable` makes the code less cluttered.

**Review by bantonsson — APPROVED** [review]

File: dd-trace-core/src/main/java/datadog/trace/core/monitor/Timer.java (hunk 2 above)
Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/core/monitor/Timer.java:43]: I looked at it, and it looks far too heavyweight for some of the things I want to get information about. It wouldn't make sense to build a histogram from from quantiles.

**Pull request merged**